src/Common/tendency.jl

```diff
@@ -513,117 +513,70 @@ end
     sb2006 = ps.rain_formation
     (; dt) = aux.TS
     (; thermo_params, common_params, air_params) = aux
-    FT = eltype(thermo_params)
     (; ρ, T) = aux.thermo_variables
     (; q_tot, q_liq, q_ice, q_rai, q_sno, N_liq, N_rai, N_aer) = aux.microph_variables
+    (; acnv, pdf_c, pdf_r, self, brek, numadj) = sb2006
     S₁ = aux.scratch.tmp
     S₂ = aux.scratch.tmp2
 
-    # helper type and wrapper to populate the tuple with sources
-    S_eltype = eltype(aux.precip_sources)
-    to_sources(args...) = S_eltype(tuple(args...))
+    # helper function to limit the tendency to the limit of the variable over the time step
+    triangle(S_x, x) = triangle_inequality_limiter(S_x, limit(x, dt))
 
     # zero out the aux sources
-    @. aux.precip_sources = to_sources(0, 0, 0, 0, 0, 0)
+    @. aux.precip_sources = CC.RecursiveApply.rzero(aux.precip_sources)
     #tot, liq, rai, Na, Nl, Nr
     if Bool(common_params.precip_sources)
         # autoconversion liquid to rain (mass)
-        @. S₁ = triangle_inequality_limiter(
-            CM2.autoconversion(sb2006.acnv, sb2006.pdf_c, q_liq, q_rai, ρ, N_liq).dq_rai_dt,
-            limit(q_liq, dt),
-        )
-        @. aux.precip_sources += to_sources(-S₁, -S₁, S₁, 0, 0, 0)
+        @. S₁ = triangle(CM2.autoconversion(acnv, pdf_c, q_liq, q_rai, ρ, N_liq).dq_rai_dt, q_liq)
+        @. aux.precip_sources.q_tot += -S₁
+        @. aux.precip_sources.q_liq += -S₁
+        @. aux.precip_sources.q_rai += S₁
 
         # autoconversion liquid to rain (number)
-        @. S₂ = triangle_inequality_limiter(
-            CM2.autoconversion(sb2006.acnv, sb2006.pdf_c, q_liq, q_rai, ρ, N_liq).dN_rai_dt,
-            limit(N_liq, dt, 2),
-        )
-        @. aux.precip_sources += to_sources(0, 0, 0, 0, -2 * S₂, S₂)
-        # liquid self_collection
-        @. S₁ =
-            -triangle_inequality_limiter(
-                -CM2.liquid_self_collection(sb2006.acnv, sb2006.pdf_c, q_liq, ρ, -2 * S₂),
-                limit(N_liq, dt),
-            )
-        @. aux.precip_sources += to_sources(0, 0, 0, 0, S₁, 0)
+        @. S₂ = triangle(CM2.autoconversion(acnv, pdf_c, q_liq, q_rai, ρ, N_liq).dN_rai_dt, N_liq)
```

Comment:
Here the limit of auto conversion tendency is `N_liq/dt/2`. To fix this you can pass `N_liq/2` instead of `N_liq` as the second argument of triangle.